diff --git a/gdk/android/gdkandroidcontentfile.c b/gdk/android/gdkandroidcontentfile.c
index 20b76185c2..c610559301 100644
--- a/gdk/android/gdkandroidcontentfile.c
+++ b/gdk/android/gdkandroidcontentfile.c
@@ -1047,7 +1047,11 @@ gdk_android_content_file_copy (GFile *file,
                                gpointer callback_data,
                                GError **error)
 {
-  g_return_val_if_fail (GDK_IS_ANDROID_CONTENT_FILE (file), FALSE);
+  if (!GDK_IS_ANDROID_CONTENT_FILE (file))
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED, "Cannot copy %s into a content file", G_OBJECT_TYPE_NAME (file));
+      return FALSE;
+    }
   if (!GDK_IS_ANDROID_CONTENT_FILE (destination))
     {
       g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED, "Cannot copy content file into %s", G_OBJECT_TYPE_NAME(destination));
